include/reactor-cpp/multiport.hh

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2022 TU Dresden
+ * All rights reserved.
+ *
+ * Authors:
+ *   Tassilo Tanneberger
+ */
+
+#ifndef REACTOR_CPP_MULTIPORT_HH
+#define REACTOR_CPP_MULTIPORT_HH
+
+#include <algorithm>
+#include <atomic>
+#include <iostream>
+#include <type_traits>
+#include <vector>
+
+namespace reactor {
+
+// fancy custom type_trait taken from stackoverflow
+// which checks if given class has the member function deactivate
+template <typename T> class has_deactivate {
+  using one = char;
+  struct two {
+    char x[2]; // NOLINT avoid-c-arrays
+  };
+
+  template <typename C> static auto test(decltype(&C::has_deactivate)) -> one;
+  template <typename C> static auto test(...) -> two;
+
+public:
+  enum { value = sizeof(test<T>(0)) == sizeof(char) };
+};
+
+// struct which gets handed to the ports to they can talk back
+// to the portbank
+struct LockedPortList {
+  std::atomic<std::size_t>* size_ = nullptr;
+  std::vector<std::size_t>* active_ports_ = nullptr;
+};
+
+template <class T, class A = std::allocator<T>>
+class Multiport { // NOLINT cppcoreguidelines-special-member-functions
+private:
+  std::vector<T> data_{};
+  std::vector<std::size_t> active_ports_{};
+  std::atomic<std::size_t> size_ = 0;
+
+public:
+  using allocator_type = A;
+  using value_type = typename A::value_type;
+  using reference = typename A::reference;
+  using const_reference = typename A::const_reference;
+  using difference_type = typename A::difference_type;
+  using size_type = typename A::size_type;
+
+  using iterator = typename std::vector<T>::iterator;
+  using const_iterator = typename std::vector<T>::const_iterator;
+
+  Multiport() noexcept = default;
+
+  ~Multiport() noexcept {
+    // we need to tell all the connected ports that this portbank is freed
+    if constexpr (std::is_pointer<T>::value) {
+      if constexpr (!has_deactivate<std::remove_pointer_t<T>>::value) {
+        for (auto i = 0; i < data_.size(); i++) {
+          data_[i]->deactivate();
+        }
+      }
+    } else {
+      if constexpr (has_deactivate<T>::value) {
+        for (auto i = 0; i < data_.size(); i++) {
+          data_[i].deactivate();
+        }
+      }
+    }
+  };
+
+  auto operator==(const Multiport& other) const noexcept -> bool {
+    return std::equal(std::begin(data_), std::end(data_), std::begin(other.data_), std::end(other.data_));
+  }
+  auto operator!=(const Multiport& other) const noexcept -> bool { return !(*this == other); };
```

Comment:
It worries me that all the API in this class is public. This would allow programmers to modify any multiport objects in reaction bodies. Could we make the none const API (and other functions that are not intended to be used in reaction bodies) private and make `T` a friend so that the ports get access?

Reply:
I don´t think this is a big problem because its the same interface that `std::vector` exposes.

Reply:
I guess you got a point there, it is not worse than before. I overlooked this when first implementing multiports. We now have the opportunity to restrict what users can do and I think we should use it :)

Reply:
will do this in extra pull request